#15  a seller should be able to mark a product as available or unavailable

oliviertech212 opened the pull request:
#### What does this PR do?

- list available products in a seller collection
- make a product available
- make a product unavailable
- document endpoint with swagger
- add translation
#### Description 
based on expiry date 
 a seller I should be able to mark a product as available or unavailable manually in order to make it available for the buyers 
#### How should this be manually tested?
- clone the repo `git clone https://github.com/atlp-rwanda/ecommerce-app-champions-bn.git`
- checkout to `ft-makeProduct-available-unavailable `
- install the packages by `npm install`
- start the server `npm run dev`
- open your browser and type ` http://localhost:{PORT}/docs`
-  first login as a vendor i.e you must be having login credentials for the vendor and validated by two-factor-auth
- if don't have an account signup as a vendor
- create product `http://localhost:{PORT}/api/product/create`
- list products you have in your database
- once you get a list of products you can use productId in searchParam to make the product. available or unavailable
Note: make sure you have Cloudinary for photo storage
### Trello cardNumber#15

sostene12 reviewed (changes requested):
You have done a great job, but it can be good if you improve this aswell.
- on the PR please clarify all endpoints to be followed to make product available or un available
- on checking the role when implementing the activity it is `re.user.role.roleName` not `req.user.roleName` change that aswell.
- you can also change that on the product.

wakaflorien commented on src/controllers/productController.js:
```diff
@@ -26,48 +26,47 @@ class productController{
+          bonus,
+          productImage
+        });
+        await product.save();
```
`await product.save();` is not needed since `create` method does it all a once on this other PR review [here](https://github.com/atlp-rwanda/ecommerce-app-champions-bn/pull/26#discussion_r1151400258)

wakaflorien commented on src/controllers/productController.js:
```diff
@@ -26,48 +26,47 @@ class productController{
-    } catch (error) {
-      return res.json({ error: error.message });
+      } catch (error) {
+        return res.json({ error: error.message });
```
- follow the same standard json response here
- also remember to put the status code e.g 

`return res.status(400).json({ status:"fail", error: error.message });

wakaflorien commented on src/controllers/productController.js:
```diff
@@ -26,48 +26,47 @@ class productController{
+    } else {
+      return res
+        .status(401)
+        .json({ status: "Oops!", message: "only suppliers allowed" });
```
- ###  status Oops! can be changed to `fail`

wakaflorien commented on src/controllers/productController.js:
```diff
@@ -163,8 +140,6 @@ class productController{
-     
-   console.log(req.user);
    if (req.user.roleName !== "vendor") {
             return res.status(401).json({
```
- responses need to be consistent 

`return res.status(401).json({
              status: "fail", message "unauthorized"});`

wakaflorien commented on src/controllers/productController.js:
```diff
@@ -255,6 +230,107 @@ class productController{
+      if (!product) {
+        return res.status(404).json({ status: "fail", message: req.t("productnotfound")});
+      }
+      return res.status(200).json({ status: req.t("success"),products:{product} });
```
- message and data attributes needs to added in the response 
- curl braces around product are not necessarly
- ` return res.status(200).json({ status: req.t("success"),message: "all products", data:product });`

wakaflorien commented on src/locales/en/translation.json:
```diff
@@ -11,7 +11,12 @@
     "profileMessage":" profile updated successfully", 
     "notFound":"profile not found",
      "Found":"profile found",
+<<<<<<< HEAD
```
- ### clearly there are some conflicting lines make sure you solve them.

wakaflorien commented on src/locales/fr/translation.json:
```diff
@@ -11,7 +11,12 @@
     "profileMessage":" Mise à jour du profil réussie", 
     "notFound":"Profil non trouvé",
      "Found":"profil trouvé",
+<<<<<<< HEAD
```
- ### make sure to resolve these conflicts.

wakaflorien commented on src/locales/kin/translation.json:
```diff
@@ -11,7 +11,12 @@
     "profileMessage":" umwirondoro wavuguruwe neza", 
     "notFound":"umwirondoro ntiwabonetse",
     "Found":"umwirondoro wabonetse",
+<<<<<<< HEAD
```
- ### resolve these too.

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    const product = await Product.findOne({ where: { id: productId } });
+    if (!product || product == null) {
+      return res
+        .status(404)
```
- redudant status codes in responses make sure you correct it.
- ` return res
        .status(404).json({ status: "fail", message: "product does not exist" });`

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: "server error" });
```
- same here avoid sending status code twice.

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    if (product) {
+      return res
+        .status(409)
+        .json({ status: 409, message: "product already exists" });
```
- same here

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: error.message });
```
- refer to above comments here too.

nicele08 commented on .env.example:
```diff
@@ -33,5 +33,12 @@ CLIENT_ID=
 JWT_SECRET=
 
+
+
```
Remove these blank lines

nicele08 commented on .env.example:
```diff
@@ -33,5 +33,12 @@ CLIENT_ID=
+api_key =
+api_secret = 
+
+
```
Same as here, we shouldn't have more than one blank line.

nicele08 commented on src/api-docs/category.docs.js:
```diff
@@ -0,0 +1,48 @@
+
+const category = {
```
This variable name is relatively clear and descriptive, but you can improve it to make it clear to other developers.
Example:
`const createCategoryDoc = {
 ....
}
`

nicele08 commented on src/controllers/productController.js:
```diff
@@ -141,7 +111,7 @@ class productController{
       }
   
-      if (req.user && req.user.role.roleName === 'vendor' && req.user.role.id !== product.vendorId) {
+      if (req.user && req.user.role.roleName === 'vendor' && req.user.role.id !== product.VendorId) {
```
You see this condition is more complex so it's a good practice to add a comment to explain the logic behind the condition.

Shumba250 committed:
rebase

Shumba250 committed:
ft-makeProduct-available
list available products in a seller collection
make a product available
make a product unavailable
document endpoint with swagger
add translation
[finishes #15 makeProduct-available]
